crates/pathfinder/src/state/sync.rs

```diff
@@ -345,32 +345,34 @@ where
 
-                    // Collect all potentially new classes.
-                    let new_root = tokio::task::block_in_place(|| {
-                        // Update state tree to determine new state root, but rollback the changes as we do
-                        // not want to persist them.
-                        let tx = db_conn
-                            .transaction_with_behavior(TransactionBehavior::Immediate)
-                            .context("Create database transaction")?;
-                        let (new_storage_commitment, new_class_commitment) = update_starknet_state(&tx, &state_update).context("Updating Starknet state")?;
-                        let new_root = StateCommitment::calculate(new_storage_commitment, new_class_commitment);
-                        tx.rollback()?;
-                        anyhow::Result::<StateCommitment>::Ok(new_root)
-                    }).context("Calculate pending state root")?;
-
-                    match new_root == state_update.new_root {
-                        true => {
-                            pending_data.set(block, state_update).await;
-                            tracing::debug!("Updated pending data");
-                        }
-                        false => {
-                            pending_data.clear().await;
-                            tracing::error!(
-                                head=%state_update.old_root,
-                                pending=%state_update.new_root,
-                                calculated=%new_root,
-                                "Pending state root mismatch"
-                            );
+                    if let Some(expected_new_root) = state_update.new_root {
+                        // Collect all potentially new classes.
+                        let new_root = tokio::task::block_in_place(|| {
+                            // Update state tree to determine new state root, but rollback the changes as we do
+                            // not want to persist them.
+                            let tx = db_conn
+                                .transaction_with_behavior(TransactionBehavior::Immediate)
+                                .context("Create database transaction")?;
+                            let (new_storage_commitment, new_class_commitment) = update_starknet_state(&tx, &state_update).context("Updating Starknet state")?;
+                            let new_root = StateCommitment::calculate(new_storage_commitment, new_class_commitment);
+                            tx.rollback()?;
+                            anyhow::Result::<StateCommitment>::Ok(new_root)
+                        }).context("Calculate pending state root")?;
+
+                        match new_root == expected_new_root {
+                            true => {
+                                pending_data.set(block, state_update).await;
+                                tracing::debug!("Updated pending data");
```

Comment:
Also, I think we should just remove the check if `new_root` matches the expected value. Care should be taken though so that we download classes even if there's no `new_root`...

Reply:
:face_exhaling: ehh I was too fast and didn't notice it's inside the block, will fix it